flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/IncrementalGroupAggregateTestPrograms.java

```diff
@@ -0,0 +1,130 @@
+    };
+
+    static final String[] SOURCE_SCHEMA = {"a INT", "b BIGINT", "c VARCHAR"};
+
+    static final TableTestProgram INCREMENTAL_GROUP_AGGREGATE_SIMPLE =
+            TableTestProgram.of(
+                            "incremental-group-aggregate-simple",
+                            "validates incremental group aggregation")
+                    .setupTableSource(
+                            SourceTestStep.newBuilder("source_t")
+                                    .addSchema(SOURCE_SCHEMA)
+                                    .producedBeforeRestore(BEFORE_DATA)
+                                    .producedAfterRestore(AFTER_DATA)
+                                    .build())
+                    .setupConfig(ExecutionConfigOptions.TABLE_EXEC_MINIBATCH_ENABLED, true)
+                    .setupConfig(
+                            ExecutionConfigOptions.TABLE_EXEC_MINIBATCH_ALLOW_LATENCY,
+                            Duration.ofSeconds(1))
+                    .setupConfig(ExecutionConfigOptions.TABLE_EXEC_MINIBATCH_SIZE, 5L)
+                    .setupConfig(
+                            OptimizerConfigOptions.TABLE_OPTIMIZER_DISTINCT_AGG_SPLIT_ENABLED, true)
+                    .setupTableSink(
+                            SinkTestStep.newBuilder("sink_t")
+                                    .addSchema("b BIGINT", "a BIGINT")
+                                    .consumedBeforeRestore("+I[1, 1]", "+I[2, 2]")
+                                    .consumedAfterRestore("-U[2, 2]", "+U[2, 3]", "+I[4, 1]")
+                                    .build())
+                    .runSql(
+                            "INSERT INTO sink_t\n"
+                                    + "     SELECT\n"
+                                    + "         b,\n"
+                                    + "         COUNT(DISTINCT a) AS a\n"
+                                    + "     FROM source_t\n"
+                                    + "     GROUP BY b")
+                    .build();
+
+    static final TableTestProgram INCREMENTAL_GROUP_AGGREGATE_COMPLEX =
+            TableTestProgram.of(
+                            "incremental-group-aggregate-complex",
+                            "validates incremental group aggregation with multiple aggregations")
+                    .setupTableSource(
+                            SourceTestStep.newBuilder("source_t")
+                                    .addSchema(SOURCE_SCHEMA)
+                                    .producedBeforeRestore(BEFORE_DATA)
+                                    .producedAfterRestore(AFTER_DATA)
+                                    .build())
+                    .setupConfig(ExecutionConfigOptions.TABLE_EXEC_MINIBATCH_ENABLED, true)
+                    .setupConfig(
+                            ExecutionConfigOptions.TABLE_EXEC_MINIBATCH_ALLOW_LATENCY,
+                            Duration.ofSeconds(10))
+                    .setupConfig(ExecutionConfigOptions.TABLE_EXEC_MINIBATCH_SIZE, 1L)
```

Comment:
@dawidwys I set the `TABLE_EXEC_MINIBATCH_SIZE` to `1` to make the results more predictable.